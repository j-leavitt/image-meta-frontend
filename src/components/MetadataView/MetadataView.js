import React from 'react';
import { Paragraph, Button, Note, CheckboxField } from '@contentful/forma-36-react-components';
import get from 'lodash/get';

async function callAPI(url) {
  console.log(`process.env.REACT_APP_API_URL`)
  const res = await fetch(`${process.env.REACT_APP_API_URL}/exif${url}`);
  const data = await res.json();
  return data.tags;
}

export class MetadataView extends React.Component {
  constructor(props) {
    super(props);

    const { fields } = props.sdk.entry;

    this.state = {
      value: "",
      overwrite: true,
      isMissingImage: !fields.image.getValue(),
      unsupportedImageType: false,
      imageRequirementsNotMet: false,
      isFetchingTags: false
    }

    // this.validateImage();
  }

  componentDidMount() {
    const { image } = this.props.sdk.entry.fields;

    image.onValueChanged(() => {
      this.setState(() => ({
        isMissingImage: !image.getValue()
      }));
      // always validate the image
      this.validateImage();
    });
  }

  validateImage = async () => {
    const { sdk } = this.props;
    const { image } = sdk.entry.fields;
    const imageId = get(image.getValue(), 'sys.id');
    if (!imageId) { return; }

    // const file = await sdk.space.getAsset(imageId);
    // const locale = sdk.locales.default;
    // const contentType = get(file, `fields.file.${locale}.contentType`);
    // const details = get(file, `fields.file.${locale}.details`);

    // test if file extension is PNG/JPEG/JPG
    // const isImageTypeValid = new RegExp(/^image\/(png|jpe?g)$/, 'i').test(contentType);
    const isImageTypeValid = true;
    // const isImageIncompatible = details.size > MAX_FILE_SIZE ||
    //                             details.width < MIN_DIMENSION_SIZE ||
    //                             details.height < MIN_DIMENSION_SIZE;
    const isImageIncompatible = false;

    this.setState(() => ({
      unsupportedImageType: !isImageTypeValid,
      imageRequirementsNotMet: isImageIncompatible
    }))
  }

  toggleOverwrite = () => {
    this.setState(state => ({
      overwrite: !state.overwrite
    }));
  }

  updateTags = async (tags) => {
    for (const [fieldName, field] of Object.entries(this.props.sdk.entry.fields)) {
      if (tags[fieldName]) {
        await field.setValue(tags[fieldName]);
      }
    }
  }

  fetchTags = async () => {
    const { sdk } = this.props;

    const imageId = get(sdk.entry.fields.image.getValue(), 'sys.id')
    const file = await sdk.space.getAsset(imageId);
    const locale = sdk.locales.default;
    const fullURL = get(file, `fields.file.${locale}.url`);
    const imagePath = new URL(`https://${fullURL}`).pathname;
    this.setState({ isFetchingTags: true });

    try {
      const exifTags = await callAPI(imagePath);

      // upload new tags
      // const newTags = this.state.overwrite ? aiTags : [...aiTags, ...this.state.tags];
      const newTags = { ...exifTags }
      this.updateTags(newTags);
    } catch (e) {
      this.props.sdk.notifier.error("Failed to obtain image metadata. Please try again later.")
    } finally {
      this.setState(() => ({ isFetchingTags: false }));
    }
  }

  updateValue = (e) => {
    this.setState({
      value: e.target.value
    });
  }

  render() {
    let hasImageError = !this.state.isMissingImage && (this.state.unsupportedImageType || this.state.imageRequirementsNotMet)
    let imageErrorMsg = this.state.unsupportedImageType ? "Unfortunately, we can only auto-tag PNG and JPG file types" : "Please make sure your image is less than 5MB and has dimensions of at least 80px for both width and height";

    return <div>
      <Paragraph>Hello Sidebar Component</Paragraph>
      {
        hasImageError &&
        <Note noteType="warning">{imageErrorMsg}</Note>
      }
      <Button
        id="fetch-tag-btn"
        buttonType="primary"
        type="button"
        disabled={this.state.isMissingImage || hasImageError}
        loading={this.state.isFetchingTags}
        onClick={this.fetchTags}
      >
        Autofill Metadata
      </Button>
      <CheckboxField
        id="overwrite-tags"
        labelText="Overwrite existing metadata"
        disabled={this.state.isMissingImage || hasImageError}
        checked={this.state.overwrite}
        onChange={this.toggleOverwrite}
      />
    </div>
  }
}

