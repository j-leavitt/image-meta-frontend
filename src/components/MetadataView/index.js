export { MetadataView } from './MetadataView.js';
