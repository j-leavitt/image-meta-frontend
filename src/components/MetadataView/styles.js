import { css } from 'emotion';
import tokens from '@contentful/forma-36-tokens';

export const styles = {
  inputWrapper: css({
    marginTop: tokens.spacingXs
  }),
  btn: css({
    marginBottom: tokens.spacingM
  }),
  fileWarning: css({
    marginTop: tokens.spacingM,
    marginBottom: tokens.spacingM
  })
}
