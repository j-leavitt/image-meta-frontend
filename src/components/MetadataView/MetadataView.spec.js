import React from 'react';
import { cleanup, render, waitFor, configure, fireEvent } from '@testing-library/react';
import fetchMock from 'fetch-mock';

import mockProps from '../../test/mockProps';
import { MetadataView } from './MetadataView';

configure({ testIdAttribute: 'data-test-id' });

const props = {
  sdk: {
    ...mockProps.sdk,
  }
};

describe('MetadataView sidebar component', () => {
  beforeEach(() => {
    props.sdk.space = {
      getAsset: jest.fn()
    };
    props.sdk.entry = {
      fields: {
        image: {
          getValue: jest.fn(),
          onValueChanged: jest.fn()
        },
        title: {
          setValue: jest.fn()
        },
        description: {
          setValue: jest.fn()
        }
      }
    };
    props.sdk.notifier = {
      error: jest.fn()
    }
  });
  afterEach(cleanup);

  it('should disable everything if theres no image', async () => {
    const appView = render(<MetadataView {...props} />);
    const { getByTestId } = appView;
    await waitFor(() => {
      expect(getByTestId('cf-ui-button').disabled).toBeTruthy();
      expect(getByTestId('cf-ui-controlled-input').disabled).toBeTruthy();
      expect(appView.container).toMatchSnapshot();
    });
  });

  it('should enable everything if theres an image', async () => {
    const imgData = {
      url: "//images.ctfassets.net/k3tebg1cbyuz/4dgP2U7BeMuk0icguS4qGw/59b8fe25285cdd1b5fcc69bd5555b3be/doge.png",
      contentType: 'image/png',
      details: { size: 200, height: 100, width: 100 }
    };
    props.sdk.space.getAsset.mockImplementation(() => ({
      fields: { file: { 'en-US': imgData } }
    }))
    props.sdk.entry.fields.image.getValue.mockImplementation(() => ({
      sys: {
        id: '098dsjnwe9ds'
      }
    }));

    const appView = render(<MetadataView {...props} />);
    const { getByTestId } = appView;
    await waitFor(() => {
      expect(getByTestId('cf-ui-button').disabled).toBeFalsy();
      expect(getByTestId('cf-ui-controlled-input').disabled).toBeFalsy();
      expect(appView.container).toMatchSnapshot();
    });
  });

  describe('Calling metadata API', () => {
    beforeEach(() => {
      const imgData = {
        url: '//images.ctfassets.net/k3tebg1cbyuz/4dgP2U7BeMuk0icguS4qGw/59b8fe25285cdd1b5fcc69bd5555b3be/doge.jpeg',
        contentType: 'image/png',
        details: { size: 200, height: 100, width: 100 }
      };
      const expectedPath = '/k3tebg1cbyuz/4dgP2U7BeMuk0icguS4qGw/59b8fe25285cdd1b5fcc69bd5555b3be/doge.jpeg'

      props.sdk.entry.fields.image.getValue.mockImplementation(() => ({
        sys: {
          id: '098dsjnwe9ds'
        }
      }));
      props.sdk.space.getAsset.mockImplementation(() => ({
        fields: {file: { 'en-US': imgData }}
      }))

      fetchMock.get(`http://localhost:4000/exif${expectedPath}`, {
        tags: {
          title: 'some_title',
          description: 'some_description'
        }
      });
    });

    afterEach(() => fetchMock.restore());

    it('should fetch tags and render them on btn click', async () => {
      const appView = render(<MetadataView {...props} />);
      const { getByTestId, getAllByTestId } = appView;

      fireEvent.click(getByTestId('cf-ui-button'));
      
      await waitFor(() => {
        expect(props.sdk.entry.fields.title.setValue).toHaveBeenCalled();
        expect(props.sdk.entry.fields.title.setValue.mock.calls[0][0]).toBe('some_title')
        expect(props.sdk.entry.fields.description.setValue).toHaveBeenCalled();
        expect(props.sdk.entry.fields.description.setValue.mock.calls[0][0]).toBe('some_description')
        expect(appView.container).toMatchSnapshot();
      });
    });
  });
});
