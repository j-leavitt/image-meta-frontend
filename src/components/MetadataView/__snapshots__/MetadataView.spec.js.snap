// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataView sidebar component Calling metadata API should fetch tags and render them on btn click 1`] = `
<div>
  <div>
    <p
      class="Paragraph__Paragraph___2aO-9"
      data-test-id="cf-ui-paragraph"
    >
      Hello Sidebar Component
    </p>
    <button
      class="Button__Button___1ZfFj a11y__focus-border--default___60AXp Button__Button--primary___JImeO"
      data-test-id="cf-ui-button"
      id="fetch-tag-btn"
      type="button"
    >
      <span
        class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
        tabindex="-1"
      >
        <span
          class="Button__Button__label___3tcOj"
        >
          Autofill Metadata
        </span>
        <svg
          class="Spinner__Spinner___32lpa Button__Button__spinner___3j8Aj Spinner__Spinner--small___2hyo0 Spinner__Spinner--white___3Hsq3 Button__Button--spinner--exit___2RUI- Button__Button--spinner-exit-active___3HXa7"
          data-test-id="cf-ui-spinner"
          viewBox="0 0 60 51"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            class="Spinner__circle1--translate___1u68A"
            transform="translate(10,33.528168)"
          >
            <g
              class="Spinner__circle1--scale___38WqL"
            >
              <circle
                id="circle1"
                r="6"
                transform="translate(0,10.471832)"
              />
            </g>
          </g>
          <g
            class="Spinner__circle2--translate___1GBUJ"
            transform="translate(30,33.528168)"
          >
            <g
              class="Spinner__circle2--scale___3khDl"
            >
              <circle
                id="circle2"
                r="6"
                transform="translate(0,10.471832)"
              />
            </g>
          </g>
          <g
            class="Spinner__circle3--translate___2M3jS"
            transform="translate(50,33.528168)"
          >
            <g
              class="Spinner__circle3--scale___2zFhX"
            >
              <circle
                id="circle3"
                r="6"
                transform="translate(0,10.471832)"
              />
            </g>
          </g>
        </svg>
      </span>
    </button>
    <div
      class="ControlledInputField__ControlledInputField___2uIG9"
      data-test-id="cf-ui-checkbox-field"
    >
      <input
        aria-label="Overwrite existing metadata"
        checked=""
        class="ControlledInput__ControlledInput___2XK3j ControlledInputField__ControlledInputField__input___3OMYB"
        data-test-id="cf-ui-controlled-input"
        id="overwrite-tags"
        type="checkbox"
        value=""
      />
      <div>
        <label
          class="FormLabel__FormLabel___3d6zQ ControlledInputField__ControlledInputField__label___a9J52"
          data-test-id="cf-ui-form-label"
          for="overwrite-tags"
        >
          Overwrite existing metadata
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`MetadataView sidebar component should disable everything if theres no image 1`] = `
<div>
  <div>
    <p
      class="Paragraph__Paragraph___2aO-9"
      data-test-id="cf-ui-paragraph"
    >
      Hello Sidebar Component
    </p>
    <button
      class="Button__Button___1ZfFj a11y__focus-border--default___60AXp Button__Button--primary___JImeO Button__Button--disabled___1E20M"
      data-test-id="cf-ui-button"
      disabled=""
      id="fetch-tag-btn"
      type="button"
    >
      <span
        class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
        tabindex="-1"
      >
        <span
          class="Button__Button__label___3tcOj"
        >
          Autofill Metadata
        </span>
      </span>
    </button>
    <div
      class="ControlledInputField__ControlledInputField___2uIG9 ControlledInputField__ControlledInputField--disabled___HDWt6"
      data-test-id="cf-ui-checkbox-field"
    >
      <input
        aria-label="Overwrite existing metadata"
        checked=""
        class="ControlledInput__ControlledInput___2XK3j ControlledInputField__ControlledInputField__input___3OMYB ControlledInput__ControlledInput--disabled___3prPF"
        data-test-id="cf-ui-controlled-input"
        disabled=""
        id="overwrite-tags"
        type="checkbox"
        value=""
      />
      <div>
        <label
          class="FormLabel__FormLabel___3d6zQ ControlledInputField__ControlledInputField__label___a9J52"
          data-test-id="cf-ui-form-label"
          for="overwrite-tags"
        >
          Overwrite existing metadata
        </label>
      </div>
    </div>
  </div>
</div>
`;

exports[`MetadataView sidebar component should enable everything if theres an image 1`] = `
<div>
  <div>
    <p
      class="Paragraph__Paragraph___2aO-9"
      data-test-id="cf-ui-paragraph"
    >
      Hello Sidebar Component
    </p>
    <button
      class="Button__Button___1ZfFj a11y__focus-border--default___60AXp Button__Button--primary___JImeO"
      data-test-id="cf-ui-button"
      id="fetch-tag-btn"
      type="button"
    >
      <span
        class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
        tabindex="-1"
      >
        <span
          class="Button__Button__label___3tcOj"
        >
          Autofill Metadata
        </span>
      </span>
    </button>
    <div
      class="ControlledInputField__ControlledInputField___2uIG9"
      data-test-id="cf-ui-checkbox-field"
    >
      <input
        aria-label="Overwrite existing metadata"
        checked=""
        class="ControlledInput__ControlledInput___2XK3j ControlledInputField__ControlledInputField__input___3OMYB"
        data-test-id="cf-ui-controlled-input"
        id="overwrite-tags"
        type="checkbox"
        value=""
      />
      <div>
        <label
          class="FormLabel__FormLabel___3d6zQ ControlledInputField__ControlledInputField__label___a9J52"
          data-test-id="cf-ui-form-label"
          for="overwrite-tags"
        >
          Overwrite existing metadata
        </label>
      </div>
    </div>
  </div>
</div>
`;
